types/rdf-validate-shacl/src/validation-report.d.ts

```diff
@@ -1,30 +1,36 @@
-import { BlankNode, DataFactory, DatasetCore, DatasetCoreFactory, NamedNode, Quad, Term } from 'rdf-js';
+import { DataFactory, DatasetCore, DatasetCoreFactory } from 'rdf-js';
 import { GraphPointer } from 'clownface';
 
+type FactoryFor<D> = D extends DatasetCore<infer OutQuad, infer InQuad> ? DataFactory<OutQuad, InQuad> & DatasetCoreFactory<OutQuad, InQuad, D> : never;
+type OutQuadOf<D> = D extends DatasetCore<infer Q, any> ? Q : never;
+type InQuadOf<D> = D extends DatasetCore<any, infer Q> ? Q : never;
+type SubjectOf<D> = OutQuadOf<D>['subject'];
+type ObjectOf<D> = OutQuadOf<D>['object'];
+
 declare namespace ValidationReport {
-    interface Options {
-        factory: DataFactory & DatasetCoreFactory;
+    interface Options<F extends DataFactory & DatasetCoreFactory> {
+        factory: F;
     }
 
-    interface ValidationResult {
-        term: BlankNode | NamedNode;
-        dataset: DatasetCore;
-        cf: GraphPointer;
-        readonly message: Term[];
-        readonly path: BlankNode | NamedNode | null;
-        readonly focusNode: BlankNode | NamedNode | null;
-        readonly severity: NamedNode | null;
-        readonly sourceConstraintComponent: BlankNode | NamedNode | null;
-        readonly sourceShape: BlankNode | NamedNode | null;
+    interface ValidationResult<D extends DatasetCore> {
+        term: SubjectOf<D>;
```

Comment:
Also note that `SubjectOf<D>` is not equivalent to `BlankNode | NamedNode`...

Same goes for `ObjectOf<>`. They both replace different kinds of types and unions